Make entire namespace selector clickable

sg00dwin opened the pull request:
Also made "Namespace" label text `pf-black-500` so that it is differentiated from the actual name.

https://jira.coreos.com/browse/CONSOLE-592?filter=-1

<img width="295" alt="screen shot 2018-07-09 at 2 58 10 pm" src="https://user-images.githubusercontent.com/1874151/42470476-8f3876d0-8388-11e8-9e64-d940d0d7eed8.png">


**before**
<img width="483" alt="screen shot 2018-07-09 at 2 41 33 pm" src="https://user-images.githubusercontent.com/1874151/42470206-c9593102-8387-11e8-83c2-549b36d8c909.png">
**after**
<img width="489" alt="screen shot 2018-07-09 at 2 41 46 pm" src="https://user-images.githubusercontent.com/1874151/42470205-c94b5a46-8387-11e8-95cb-6e7989720c31.png">
**before**
<img width="561" alt="screen shot 2018-07-09 at 2 40 59 pm" src="https://user-images.githubusercontent.com/1874151/42470207-c9678e96-8387-11e8-97af-d2cae3cced1f.png">
**after**
<img width="555" alt="screen shot 2018-07-09 at 2 40 22 pm" src="https://user-images.githubusercontent.com/1874151/42470208-c973cd82-8387-11e8-9999-f84298ddaeb0.png">


---

![namespace-selector4](https://user-images.githubusercontent.com/1874151/42470277-f8df2cb0-8387-11e8-8b32-49a75d2a2741.gif)

spadgett commented on frontend/public/components/_dropdown.scss:
```diff
@@ -91,29 +91,40 @@ $color-bookmarker: #DDD;
+
+    &:before {
+      color: $color-pf-black-500;
+      content: 'Namespace:';
```
The label needs to change depending on whether this is OpenShift or vanilla Kubernetes.

spadgett commented on frontend/public/components/_dropdown.scss:
```diff
@@ -91,29 +91,40 @@ $color-bookmarker: #DDD;
+    }
+
+    &:before {
+      color: $color-pf-black-500;
```
Can't decide if I like the color :/

@openshift/team-ux-review ?

sg00dwin replied:
To me it helps set apart the label from the name.

tlwu2013 replied:
I feel the color contrast between the text and the bg-color is a bit too low, so did a quick WCAG check.
Turned out the original color meet the guideline but not the new one..
**original color: pf-black-300 (#d1d1d1)**
![color_ori_ d1d1d1](https://user-images.githubusercontent.com/5903705/42473514-b13747de-8379-11e8-9cc4-090aab6659b0.png)

**new color: pf-black-500 (#8b8d8f)**
![color_new_ 8b8d8f](https://user-images.githubusercontent.com/5903705/42473504-aae1aeba-8379-11e8-8508-4d32940912fb.png)

Since pf-black-500 is too dark and pf-black-400 is almost no difference here, should we just keep the same color (pf-black-300) here?   Thoughts? @cshinn , @beanh66

tlwu2013 replied:
Other than the color change, I think make the entire namespace selector clickable is awesome! Long waiting change!

rhamilto replied:
> Can't decide if I like the color :/

My initial impression was the same.  My vote is to keep it the same given @tlwu2013's concern regarding contrast.

benjaminapetersen replied:
Contrast is a great Q.  I think the subtle difference is helpful, but not essential.  If it was something essential, I think it would be critical to pass an accessibility test. Since its a subtle difference to slightly improve readability, not sure it needs to pass the same test.

Curious if there are accessibility guidelines for things that are "subtle but not essential".  I'm guessing its probably not quite that sophisticated yet.

benjaminapetersen replied:
Btw, I may have misunderstood.

- If it is `contrast is not high enough against background color`, this is a big deal.
- If it is `contrast is not significant enough between the two text colors`, this is probably not a big deal.

Thats what I meant in my previous comment.

spadgett commented:
I notice in OpenShift console, we didn't show a label on the dropdown. But I think it's helpful to have it.

sg00dwin commented:
@spadgett @rhamilto 
moved Namespace: back into markup and reverted color.

**Clickable area**
<img width="372" alt="screen shot 2018-07-10 at 1 27 51 pm" src="https://user-images.githubusercontent.com/1874151/42526898-2d0ec118-8445-11e8-92bd-b32065ee9478.png">

---
<img width="506" alt="screen shot 2018-07-10 at 12 52 44 pm" src="https://user-images.githubusercontent.com/1874151/42526970-58d79fae-8445-11e8-9734-7b2c03b37b3c.png">

<img width="332" alt="screen shot 2018-07-10 at 12 52 28 pm" src="https://user-images.githubusercontent.com/1874151/42526800-df9b1120-8444-11e8-873c-d2c8ded27715.png">

spadgett commented on frontend/public/components/namespace.jsx:
```diff
@@ -247,14 +247,14 @@ class NamespaceDropdown_ extends React.Component {
         canFavorite
         items={items}
         title={title}
+        labelPrefix={model.label}
```
Why not just make this part of title if we have no special styling?

```jsx
title={`${model.label}: ${title}`}
```

sg00dwin replied:
Well, I did add a small amount of extra spacing between the colon and name, which I think helps the name standout better, but will change...

<img width="275" alt="screen shot 2018-07-11 at 10 57 38 am" src="https://user-images.githubusercontent.com/1874151/42580973-bece9c40-84f9-11e8-94e3-e33286dcf1e9.png">
<img width="268" alt="screen shot 2018-07-11 at 11 03 22 am" src="https://user-images.githubusercontent.com/1874151/42581129-110edccc-84fa-11e8-857c-ee00f4499013.png">

sg00dwin replied:
pr rebased and updated

spadgett commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -287,10 +287,11 @@ export class Dropdown extends DropdownMixin {
 
     const button = noButton
       ? <div onClick={this.toggle} className="dropdown__not-btn" id={this.props.id}>
+        {labelPrefix && `${labelPrefix}: ` }
```
Remove the changes to dropdown.jsx?

spadgett commented on frontend/public/components/namespace.jsx:
```diff
@@ -247,14 +247,13 @@ class NamespaceDropdown_ extends React.Component {
         canFavorite
         items={items}
-        title={title}
+        title={`${model.label}:   ${title}`}
```
I think the extra spaces will be collapsed unless we add a `white-space` style or maybe HTML whitespace entities.

We could add a titlePrefix prop if we want to style it differently, but in that case we'd want to handle the button case. Your original fix only added it for noButton.

sg00dwin commented:
/retest

spadgett commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -287,7 +287,7 @@ export class Dropdown extends DropdownMixin {
   render() {
     const {active, autocompleteText, selectedKey, items, title, bookmarks, keyboardHoverKey, favoriteKey} = this.state;
-    const {autocompleteFilter, autocompletePlaceholder, noButton, noSelection, className, menuClassName, storageKey, canFavorite, dropDownClassName} = this.props;
+    const {autocompleteFilter, autocompletePlaceholder, noButton, noSelection, className, menuClassName, storageKey, canFavorite, dropDownClassName, labelPrefix} = this.props;
```
You'll want to undo this change, too. (That's probably why tests are failing.)

sg00dwin commented:
I think she's finally ready. /me crosses fingers

spadgett commented:
Test failure looks related:

```
1) Kubernetes resource CRUD operations Pod : has a working namespace dropdown on namespaced objects
   Error: Failed expectation
       at Object.<anonymous> /go/src/github.com/openshift/console/frontend/integration-tests/tests/crud.scenario.ts:79:61
       at step /go/src/github.com/openshift/console/frontend/integration-tests/tests/crud.scenario.ts:33:23
       at Object.next /go/src/github.com/openshift/console/frontend/integration-tests/tests/crud.scenario.ts:14:53
       at fulfilled /go/src/github.com/openshift/console/frontend/integration-tests/tests/crud.scenario.ts:5:58
       at process._tickCallback internal/process/next_tick.js:68:7
```

sg00dwin commented:
Well, looking at the test https://github.com/openshift/console/blob/master/frontend/integration-tests/tests/crud.scenario.ts I would guess that it's looking for an {Exact_name} and instead it's {Namespace: Exact_name}  But I'm going to need guidance on what/where to change.

benjaminapetersen commented:
@sg00dwin I'll ping you in slack w/some info on debugging the test.

sg00dwin commented:
@spadgett PTAL
